### ISSUE TITLE
BlocUnhandledErrorException in assert does not get fired in release mode

### ISSUE THREAD

**Issue by magicleon94** [felangel/bloc#2278]
Hi,

I was having some issues with my flutter app because of an exception not being logged on sentry when in release mode, so I did come digging in the `flutter_bloc` code and found that the `BlocUnhandledErrorException` is thrown only in debug mode [because it is wrapped in an assert](https://github.com/felangel/bloc/blob/97f826273182f738a14224ca670ebc96c3045e6d/packages/bloc/lib/src/bloc.dart#L388).

I'm not sure if it's a desired behavior or not, but I think that letting this throw in release mode too would improve unhandled exceptions logging on a release app, or a release-compiled dart code in general, I guess, since this is not `flutter_bloc`.

If this is a desired behavior, may I ask why? And most importantly, my workaround was to "catch" the error via a `BlocObserver` and reporting the error to sentry from there... is there a better solution?

Thank you!

Antonello

**Comment by felangel** [comment]
Hi @magicleon94 👋 
Thanks for opening an issue!

This is working as intended. In release mode, errors still get reported to BlocObserver but they do not get treated as unhandled exceptions in order to ensure that the app does not crash.

>my workaround was to "catch" the error via a BlocObserver and reporting the error to sentry from there 

Can you share some sample code to illustrate what you mean? Thanks!

**Comment by magicleon94** [comment]
Hi @felangel !
Thank you for the quick reply :)

This is what I mean

```dart
// inside BlocObserver
@override
  void onError(Cubit cubit, Object error, StackTrace stackTrace) {
    log('onError -- cubit: ${cubit.runtimeType}, error: $error');
    if (kReleaseMode) {
      Catcher.reportCheckedError(error, stackTrace);
    }
    super.onError(cubit, error, stackTrace);
  }
```

As you said, on release mode the errors are reported but not thrown, so I just report to Sentry (via Catcher) when the observer reports the error :)

The code is for an older version of `bloc`(it should be `BlocBase` instead of `Cubit`) but I didn't update the dependency yet, but the "issue" is still valid.

Good to know that this is an intended behavior, it wasn't clear in the beginning and it felt strange (to me, at least). Is there something on the docs that I might have missed? Is it worth documenting this?

Thank you!

**Comment by felangel** [comment]
@magicleon94 thanks for clarifying! I would recommend having an error reporter or logger interface with different concrete implementations based on whether the app is running in release mode. Then you can inject the reporter into the `BlocObserver` and it can just report the error without needing the logic of checking whether `kReleaseMode` is true.

```dart
abstract Logger {
  void reportError(Object error, StackTrace stackTrace);
}

class LoggerA implements Logger {
  @override
  void reportError(Object error, StackTrace stackTrace) {
    // report error
  }
}

class LoggerB implements Logger {
  @override
  void reportError(Object error, StackTrace stackTrace) {
    // report error
  }
}

void main() {
  final logger = kReleaseMode ? LoggerA() : LoggerB();
  Bloc.observer = BlocObserver(logger: logger);
  ...
}
```

I'll keep this issue open to update the docs 👍

**Comment by magicleon94** [comment]
Hi and thank you, 
this is an approach I considered but due to the circumstances I didn't consider it worth pursuing, since I already have different "implementations" of `Catcher` based on the build mode (more a different composition than different implementations).
The if there was just to avoid writing on console two times 🙈

Glad of having been of some kind of help, cheers!

**Comment by felangel** [comment]
This is closed by #2899 and is included in bloc 8.0.0-dev.4
